doc/src/build/dependency-overrides.md

```diff
@@ -0,0 +1,57 @@
+
+```move
+[package]
+name = "vault"
+version = "1.0.0"
+
+[dependencies]
+crypto = { git = "https://github.com/crypto.org/crypto.git" , rev = "v1.0.0"}
+```
+
+```move
+[package]
+name = "currency"
+version = "2.0.0"
+
+[dependencies]
+crypto = { git = "https://github.com/crypto.org/crypto.git" , rev = "v2.0.0"}
+```
+
+This situation represents the [_diamond dependency conflict_](https://jlbp.dev/what-is-a-diamond-dependency-conflict) where two indirect dependencies in turn depend on different versions of the same package. If all packages involved are developed independently, then the developer of the `user` package has a problem -- the Move compiler cannot discern which version of the `crypto` package to use and, as a result, building of the `user` package fails.
+
+To resolve these types of problems, developers can _override_ indirect dependencies of the packages they develop. In other words, we provide a mechanism that the developer can use to specify a single version of a dependent package to be used during the build process. This can be done by introducing a _dependency override_ in the manifest file of the developed package. In the case of our running example, developer of the `user` package might decide that all dependent packages should use version `2.0.0` of the `crypto` package (note that it is the same as specified in the manifest file of the `currency` package but could be a version not used by any of the dependencies, for example `3.0.0`). At a technical level, in order to enforce a dependency override, a developer of a "user" package  has to put a dependency with a chosen version in their own manifest file, with the addition of the `override` flag set to `true`. The manifest file of the `user` package correcting the diamond dependency conflict in our running example would resemble the following:
+
+```move
+[package]
+name = "user"
+version = "1.0.0"
+
+[dependencies]
+vault = { git = "https://github.com/vault_org/vault.git" }
+currency = { git = "https://github.com/currency_org/currency.git" }
+crypto = { git = "https://github.com/crypto.org/crypto.git" , rev = "v2.0.0" , override = true }
+```
+
+Note that overriding transitive dependencies in such a manner might not always be successful. In particular, the `vault` package in our example might truly depend on the functionality of the `crypto` package that is only available in version `1.0.0` of this package (e.g, was removed in later package revisions), in which case the build would also be unsuccessful. By offering the ability to enforce dependency overrides we merely put another tool into the developer's toolbox in the hope that it would allow resolution of transitive dependency conflicts in the majority of cases.
```

Comment:
It doesn't have to be in this PR, but we should also explain the on-chain constraint around versioning and compatibility, and relatedly, the lack of a relationship between on-chain and local versions.  I.e. in practice, if folks are using semantic versioning, then `v2.0.0` is not going to be compatible with `v1.0.0` and so it won't be a package upgrade in the technical sense, and you won't be able to perform the override, and they may choose an override that makes everything compile, but again find themelves in hot water because it's not a strictly greater version on-chain.

This is another place where Automatic Address Management (or an extension of it) will help.

Reply:
I agree! And I tried to think what the best place for the additional information would be and I am currently leaning towards a separate "page", perhaps titled "Package Versioning" (or "Code Versioning" to avoid title overlap with "Package Upgrades"). I will start writing it soon-ish (unless someone preempts me, but then please let me know :-) ), but will likely as @tzakian to fill in the compatibility part. If we decide to put it with "Dependency Overrides" after all, we can indeed always augment this document in a separate PR.